### PR TITLE
2d fluid flow on car mesh

### PR DESCRIPTION
**Pull request by flothesof**
Hi,

as discussed in https://github.com/sfepy/sfepy/issues/743, here is a PR to add this example to sfepy.

I have a few questions:
- is it possible to tell resview how to add streamlines to the plot? I wrote a pyvista script to animate this example, which could also be part of the file (see results here https://twitter.com/frolianlb/status/1460231767542190083)
- should I add the .geo script to the repo? if so, how do I document its existence in the text?
- is the mesh too big for this kind of example? (2MB)

Cheers,
Florian

### CODE REVIEW

**Comment by rc** [comment]
Hi Florian,

thanks, looks good! I have only some remarks below.

> I have a few questions:
> 
>     * is it possible to tell resview how to add streamlines to the plot? I wrote a pyvista script to animate this example, which could also be part of the file (see results here https://twitter.com/frolianlb/status/1460231767542190083)

Nice animation! IMO resview cannot add streamlines now (cc-ing @vlukes), but adding the visualization code to the example would work well.
 
>     * should I add the .geo script to the repo? if so, how do I document its existence in the text?

You can document  the mesh generation in the docstring.

>     * is the mesh too big for this kind of example? (2MB)

It is OK,  but if you could get reasonable results with a smaller mesh it might be better.
 
Now the remarks:

- Make sure, please, that also the lines in the docstring are no more than 80 characters long.
- Consider using `\nabla` for the gradient and `\nabla \cdot` for the divergence like in the other examples.
- The second dosctring equation has missing `}`, use:
```
    \int_{\Omega} \nabla \psi \cdot \nabla \phi
    = \int_{\Gamma_{left}} \ul{v}_0 \cdot n \psi
    + \int_{\Gamma_{right}} \ul{v}_0 \cdot n \psi
    \;, \quad \forall \psi \;,

where :math:`\ul{v}_0` is ....
```
and define v_0.
- There are no Dirichlet BCs, so the matrix is singular. The `'ls.scipy_direct'` (UMFPACK) solver is robust so it only complains, but other solvers might simply fail. Try fixing the potential in a single vertex, for example:
```
regions = {
    'Omega' : 'all',
    'Gamma_Left' : ('vertices in (x < 0.1)', 'facet'),
    'Gamma_Right' : ('vertices in (x > 1919.9)', 'facet'),
    'Vertex' : ('vertex in r.Gamma_Left', 'vertex'),
}
...
ebcs = {
    'fix': ('Vertex', {'phi.0': 0.0}),
}
```

**Comment by flothesof** [comment]
I’ve slightly extended the example (possibility of taking into account any far field v0) and rewritten the equations as you suggested. 

@vlukes do you have any thoughts about the streamlines? they are available in pyvista so maybe this could be integrated into resview.py?

File: examples/diffusion/laplace_fluid_2d.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+"""
+from sfepy import data_dir
+
+filename_mesh = data_dir + '/meshes/2d/citroen.h5'
+
+v0 = np.array([1, 0.25])
+
+materials = {
+    'm': ({'sigma': 1.0},),
+    'f': ({'left': np.dot(v0.T, np.array([-1, 0])),
```
Comment (rc) [inline examples/diffusion/laplace_fluid_2d.py:57]: NumPy import is missing. (Also consider using `nm` like most other examples do (historical reasons)).

File: examples/diffusion/laplace_fluid_2d.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+integrals = {
+    'i': 2,
+}
+
+equations = {
+    'Laplace equation':
+        """dw_laplace.i.Omega( m.sigma, psi, phi )
+         = dw_integrate.i.Gamma_Left(f.left, psi)
+         + dw_integrate.i.Gamma_Right(f.right, psi)"""
```
Comment (rc) [inline examples/diffusion/laplace_fluid_2d.py:94]: The top and bottom regions and material are not used?

File: examples/diffusion/laplace_fluid_2d.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    python3 resview.py citroen.vtk
+
+
+Generating the mesh
+-------------------
+
+The mesh can be generated with::
+    gmsh -2 -f msh22 meshes/2d/citroen.geo -o meshes/2d/citroen.msh
+    python script/convert_mesh.py --2d meshes/2d/citroen.msh \
+        meshes/2d/citroen.h5
```
Comment (rc) [inline examples/diffusion/laplace_fluid_2d.py:47]: Use  `python3`. It is also OK to have the command line examples longer than 80 chars - a single line is easier to copy & paste.

**Comment by rc** [comment]
Thanks for the update! I have realized that you might want to use `dw_surface_ndot` term - it would multiply by the normal automatically.

**Comment by vlukes** [comment]
> @vlukes do you have any thoughts about the streamlines? they are available in pyvista so maybe this could be integrated into resview.py?

Hi @flothesof! Yes, I will try to implement streamlines into `resview.py`.

File: examples/diffusion/laplace_fluid_2d.py (hunk 1 above)
Comment (flothesof) [reply examples/diffusion/laplace_fluid_2d.py:57]: I’ll add it (I forgot to re-test after copying from my main script)

File: examples/diffusion/laplace_fluid_2d.py (hunk 2 above)
Comment (flothesof) [reply examples/diffusion/laplace_fluid_2d.py:94]: Please see my next version (I forgot to re-test after copying from my main script) where the domains are used :)

File: examples/diffusion/laplace_fluid_2d.py (hunk 3 above)
Comment (flothesof) [reply examples/diffusion/laplace_fluid_2d.py:47]: I’ll make a longer line then!

**Comment by rc** [comment]
Thanks @flothesof, almost there! Just two(three) things to settle:
- Use `python3` as the executable name in the docstring (line 46).
- Note that you can use `dw_surface_ndot` term - using it would make it easier for the normals not aligned with coordinate axes.
- We can either wait to have the streamlines in resview (thanks @vlukes!), or merge the example as it is and update it later.

**Comment by vlukes** [comment]
> We can either wait to have the streamlines in resview (thanks @vlukes!), or merge the example as it is and update it later.

I suggest to merge the example as it is. I will implement the streamlines to `resview` later.

**Comment by flothesof** [comment]
I used the `dw_surface_ndot ` term, as far as I could tell it there wasn’t any example using it so far.
Executing the script should look like this:
image

**Comment by rc** [comment]
> I used the `dw_surface_ndot ` term, as far as I could tell it there wasn’t any example using it so far. @rc are the examples automatically linked to the terms table page? I find it very useful, as a beginner user of Sfepy, to be able to look at an example featuring the terms I’m trying to use.

Yes, the links are automatically generated (PR #374). They are meant exactly for this purpose :)

Thanks for the update, LGTM! I will merge it now, and later we can add the instructions how to display the streamlines.

Maybe attach your visualization script to this PR, so that we can take inspiration for the resview update?

**Comment by rc** [comment]
Thanks for your contribution!

**Comment by flothesof** [comment]
You’re welcome :)

**Comment by flothesof** [comment]
Here’s a sample animation script that I have used (ping @vlukes for streamlines computed with the gradient):
```python
# load the mesh
import pyvista as pv
import numpy as np

mesh = pv.read("citroen.vtk")

# add gradients
grad = mesh.compute_derivative(scalars="phi")

# get min max coords of mesh
minimaxi = np.c_[mesh.points.min(axis=0), mesh.points.max(axis=0)][:2, :]


def make_points(minimaxi, n_points=100):
    points = np.zeros((n_points, 3))
    for coord in range(2):
        low, high = minimaxi[coord, 0], minimaxi[coord, 1]
        points[:, coord] = np.random.randint(low=low, high=high, size=n_points)
    return points

# and now an animation
points = make_points(minimaxi, n_points=10000)
point_cloud = pv.PolyData(points)
point_cloud = point_cloud.interpolate(grad, null_value=0, radius=20.)

streamlines = grad.streamlines(
    pointa=(0.1, 0, 0),
    pointb=(0.1, minimaxi[1, 1], 0),
    n_points=20,
    max_time=2000.,
    compute_vorticity=False,  
)


plotter = pv.Plotter()
plotter.add_mesh(grad, show_scalar_bar=False)
plotter.add_mesh(streamlines, show_scalar_bar=False)
plotter.add_mesh(point_cloud, scalars="phi", show_scalar_bar=False, colormap="inferno", opacity=0.5)
plotter.view_xy()
plotter.camera_position = [(868.4118785064393, 500.3989374701676, 2417.938072036079),
 (868.4118785064393, 500.3989374701676, 50.0),
 (0.0, 1.0, 0.0)]
plotter.open_movie("citroen.mp4")

for i in range(500):
    print(i, interpolated['gradient'].min(), interpolated['gradient'].max())
    interpolated = point_cloud.interpolate(grad, null_value=0, radius=20, strategy='mask_points')
    points = interpolated.points + 10.5 * interpolated["gradient"]
    points[:, 0] = points[:, 0] % minimaxi[0, 1]
    points[:, 1] = points[:, 1] % minimaxi[1, 1]
    plotter.update_coordinates(points, render=False)
    point_cloud = pv.PolyData(points)
    plotter.render()
    plotter.write_frame()
plotter.show()
plotter.close()
```

**Pull request merged**